### PR TITLE
Feature/208 provide info about dce runs

### PR DESCRIPTION
**Pull request by DzMakatun**
Adds Monitoring tab to Dataset page, where a user can select INFO DATE [run->controlMeasure->metadata->informationDate] interval and see visualization of job and record statuses.  This view focuses on the chronology and state of the uploaded data (rather than runs alone). The idea is to let users check if the data for the dates of their interest were processed correctly, regardless of when exactly the processing occurred.

Comments are appreciated.

### CODE REVIEW

**Comment by Zejnilovic** [comment]
## 1 No data shown
I have a run that I just executed and I cannot see it.
image

**Comment by DzMakatun** [comment]
> ## 1 No data shown
> I have a run that I just executed and I cannot see it.
> image

Please, note that the runs are filtered by the INFO DATE [run->controlMeasure->metadata->informationDate] (not the job execution timestamps). This view focuses on the chronology and state of the uploaded data (rather than runs alone). The idea is to let users check if the data were processed correctly for the dates of interest.

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+
+
+  def getMonitoringDataPoints(datasetName: String, startDate: String, endDate: String): Future[Seq[String]] = {
+    // scala mongodb driver does not yet support all mql features, so we use Document() with aggregate pipelines
+    val observable: AggregateObservable[Document] = collection
+      .aggregate(Seq(
+        // filter by dataset name
+        filter(equal("dataset", datasetName)),
+        //add casted dates
+        Document("""{$addFields: {
```
Comment (lokm01) [inline menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala]: Any chance we could use the scala DSL for these?

File: menas/src/main/scala/za/co/absa/enceladus/menas/services/MonitoringService.scala [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import scala.concurrent.Future
+
+@Service
+class MonitoringService @Autowired()(monitoringMongoRepository: MonitoringMongoRepository)
+  extends ModelService(monitoringMongoRepository) {
+
+  import scala.concurrent.ExecutionContext.Implicits.global
+
+  @Value("${za.co.absa.enceladus.spline.urlTemplate}")
+  val splineUrlTemplate: String = null
```
[the remark this answers is not in the document]
Comment (lokm01) [reply menas/src/main/scala/za/co/absa/enceladus/menas/services/MonitoringService.scala]: Where is this being used anyway?

File: menas/ui/components/Component.js [hunk 3]
```diff
@@ -14,7 +14,7 @@
  */
 
 sap.ui.define([
-    "sap/ui/core/UIComponent"
+  "sap/ui/core/UIComponent",
```
Comment (lokm01) [inline menas/ui/components/Component.js]: There's only one element in that array - why the comma?

File: menas/ui/components/dataset/datasetDetail.controller.js [hunk 4]
```diff
@@ -39,6 +40,18 @@ sap.ui.define([
       let cont = new ConformanceRuleDialog(this);
       let view = this.getView();
 
+      view.byId("monitoring-multiheader-info").setHeaderSpan([2,1]);
```
Comment (lokm01) [inline menas/ui/components/dataset/datasetDetail.controller.js]: Can these configs be specified through the view?

File: menas/ui/components/dataset/datasetDetail.controller.js [hunk 5]
```diff
@@ -39,6 +40,18 @@ sap.ui.define([
...
+      view.byId("monitoring-multiheader-checkpoint").setHeaderSpan([3,1]);
+      view.byId("monitoring-multiheader-recordcount-raw").setHeaderSpan([5,1,1]);
+      view.byId("monitoring-multiheader-recordcount-std").setHeaderSpan([0,2,1]);
+      view.byId("monitoring-multiheader-recordcount-cnfrm").setHeaderSpan([0,2,1]);
+
+      this.oFormatYyyymmdd = sap.ui.core.format.DateFormat.getInstance({
```
Comment (lokm01) [inline menas/ui/components/dataset/datasetDetail.controller.js]: Would Formatters be better place for this?

File: menas/ui/service/MonitoringService.js [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+      estimatedRecordCount = (+oRun["conform_records_succeeded"]) + (+oRun["conform_records_failed"] )
+    }
+    recordsStatusAggregator["Inconsistent info"].recordCounts.push(estimatedRecordCount)
+    recordsStatusAggregator["Inconsistent info"].counter += estimatedRecordCount
+  };
+
+
+  this.getData= function(sId, sStartDate, sEndDate) {
+    MonitoringService.clearAggregators();
+    Functions.ajax("api/monitoring/data/datasets/"
```
Comment (lokm01) [inline menas/ui/service/MonitoringService.js]: I think this is the left-over way from the earlier components. 

Could you use the RestDAO.js for these calls? It'd be great to unify on one.

**Comment by Zejnilovic** [comment]
## 2 Redrawing of graphs
Graphs get redrawn on every click to the calendar. There should be a possibility of distinguishing between drag and drop motion and two clicks. If not, I would only keep the 2 click option.

Jun-10-2019 07-51-33

**Comment by Zejnilovic** [comment]
## 3 Summary date format
I would format the date in a more standard way of "%dd/%mm/%yyyy" for easier user experience.

image

**Comment by Zejnilovic** [comment]
## 4 Make Records per run absolute values
So if I have thousands of records I cannot see myself caring if I ingested 10K or 10K + 1 values. But I will care to see how the run does in comparison to others and the human eye cannot distinguish this if this is not put into absolute values.

**Comment by Zejnilovic** [comment]
## 5 Button under headers
The is a button under some headers and none under others. The button in general controls the view of the whole table, which makes it confusing when it is localized to one column, even more, when there are subcolumns named the same.

Jun-10-2019 10-56-28

Could this functionality be moved to some dropdown or something above the table?

**Comment by Zejnilovic** [comment]
## 6 Size of the column graph
Column graph size greatly differs depending on the number of records in the selections. This could be unreadable with more records. I would set the record limit of ~20 (??) and in case there is more just show a warning, that there are more of them but showing the just the 20. 

Or maybe let the user choose how many are displayed and when he displays 100 of them and it is hard to see, it is not our issue.

**Comment by Zejnilovic** [comment]
## 7 Unable to do cross month selection
I can imagine a use case where I would need to get a few days from the first month and a few from a second. Currently, I am unable to do so.

Jun-10-2019 12-55-14

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.menas.repositories
+
+import org.mongodb.scala._
+//import org.mongodb.scala.{AggregateObservable, Completed, Document, MapReduceObservable, MongoDatabase}
+//import org.mongodb.scala.bson.BsonDocument
+//import org.mongodb.scala.bson.conversions.Bson
```
[the remark this answers is not in the document]
Comment (DzMakatun) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala]: ok

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+                   |          latestCheckpoint: {$first: {$slice: ["$controlMeasure.checkpoints", -1]}},
+                   |          "publish_dir_size" : {$first: "$controlMeasure.metadata.additionalInfo.publish_dir_size"},
+                   |          "std_dir_size" : {$first: "$controlMeasure.metadata.additionalInfo.std_dir_size"},
+                   |          "raw_dir_size" : {$first: "$controlMeasure.metadata.additionalInfo.raw_dir_size"}
+                   |
+                   |      }}""".stripMargin),
+        // sort the final results
+        Document("""{$sort: {informationDateCasted : -1, reportVersion: -1}}""".stripMargin)
+        //Document("""{$limit: 20}""".stripMargin)
+
```
[the remark this answers is not in the document]
Comment (DzMakatun) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala]: ok

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (DzMakatun) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala]: #208 Clean up according to the comments

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (DzMakatun) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala]: #208 Clean up according to the comments

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+      ))
+
+    //val observable: AggregateObservable[Document] = collection.aggregate(Seq(filter(equal("dataset", datasetName))))
```
[the remark this answers is not in the document]
Comment (DzMakatun) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala]: ok, removed #208 Clean up according to the comments

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+
+    observable.map(doc => doc.toJson).toFuture()
+  }
+
+  // TODO: monitoring of the latest events (checkpoints) in the system
```
[the remark this answers is not in the document]
Comment (DzMakatun) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala]: ok, not likely to be used, removed #208 Clean up according to the comments

File: menas/src/main/scala/za/co/absa/enceladus/menas/services/MonitoringService.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (DzMakatun) [reply menas/src/main/scala/za/co/absa/enceladus/menas/services/MonitoringService.scala]: thanks for spotting, removed #208 Clean up according to the comments

File: menas/ui/components/dataset/datasetDetail.controller.js [hunk 11]
```diff
@@ -237,5 +257,67 @@ sap.ui.define([
...
+
+    setDefaultMonitoringDateInterval: function () {
+      let oEnd = new Date();
+      let oStart = new Date(oEnd.getTime() - 14 * 24 * 60 * 60 * 1000); // Two weeks before today
+      let oCalendar = this.byId("calendar");
+      oCalendar.removeAllSelectedDates();
+      oCalendar.addSelectedDate(new sap.ui.unified.DateRange({startDate: oStart, endDate: oEnd}))
+      this._model.setProperty("/monitoringDateFrom", this.oFormatYyyymmdd.format(oStart))
+      this._model.setProperty("/monitoringDateTo", this.oFormatYyyymmdd.format(oEnd))
+      //this._updateTimeInterval(oCalendar.getSelectedDates()[0]);
```
[the remark this answers is not in the document]
Comment (DzMakatun) [reply menas/ui/components/dataset/datasetDetail.controller.js]: thnx, #208 Clean up according to the comments

File: menas/ui/components/external/it/designfuture/chartjs/library-preload.js [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
...configuration is at Chart.defaults.global.hover.\r\n\t\t\t\t///////////////////////////////////////////////////////////////////////////////////\r\n\t\t\t\t\r\n\t\t\t\t/**\r\n\t\t\t\t * Sets which elements appear in the tooltip.\r\n\t\t\t\t */\r\n\t\t\t\thoverMode : {type : \"string\", group : \"Appearance\", defaultValue : \"label\"},\r\n\t\t\t\t\r\n\t\t\t},\r\n\t\t\taggregations: {},\r\n\t\t\tevents: {}\r\n\t\t},\r\n\t\t\r\n\t\tonBeforeRendering: function() {\r\n\t\t\tthis.setChartType(this.getIsHorizontal() ? \"horizontalBar\" : \"bar\");\r\n\t\t\tBaseChartJS.prototype.setHoverMode.call(this, this.getHoverMode());\r\n\t\t},\r\n\t\t\r\n\t\t/////////////////////////////////////\r\n\t\t// OVERRIDE\r\n\t\t/////////////////////////////////////\r\n\t\t\r\n\t\taddGlobalOptions: function(globalOptions) {\r\n\t\t},\r\n\t\t\r\n\t\t//////////////////////////////////////////////\r\n\t\t// GETTER / SETTER\r\n\t\t//////////////////////////////////////////////\r\n\t\t\r\n\t\t\r\n\t\t\r\n\t\t//////////////////////////////////////////////\r\n\t\t// CHART METHODS\r\n\t\t//////////////////////////////////////////////\r\n\t\t\r\n\t});\r\n\r\n\r\n\treturn BarChartJS;\r\n\r\n}, /* bExport= */ true);",
```
[the remark this answers is not in the document]
Comment (DzMakatun) [reply menas/ui/components/external/it/designfuture/chartjs/library-preload.js]: "openui5-chartjs is a custom OpenUI5 library that wraps Chart.js. SAPUI5 already offer a chart library called VizFrame, but it does not come for free and it's a little bit outdated if compared to rest of others web framework/library."

File: menas/ui/components/Component.js (hunk 3 above)
Comment (DzMakatun) [reply menas/ui/components/Component.js]: thnx, #208 Clean up according to the comments

File: menas/ui/components/dataset/datasetDetail.controller.js (hunk 4 above)
Comment (DzMakatun) [reply menas/ui/components/dataset/datasetDetail.controller.js]: Unfortunately not

File: menas/ui/components/dataset/datasetDetail.controller.js (hunk 5 above)
Comment (DzMakatun) [reply menas/ui/components/dataset/datasetDetail.controller.js]: sure, thnx #208 Clean up according to the comments

**Comment by DzMakatun** [comment]
> ## 3 Summary date format
> I would format the date in a more standard way of "%dd/%mm/%yyyy" for easier user experience.
> 
> image

The yyyy-MM-dd format matches the infodate (in the run collection, directory tree, job parameters,  and chart labels). I would keep it this way, to be uniform across the monitoring data.

**Comment by DzMakatun** [comment]
> cross month selection
Opened a task to improve interval selection #459.
For now,  you can navigate to the next/previous month with the drag-n-drop motion only.

**Comment by DzMakatun** [comment]
> ## 2 Redrawing of graphs
> Graphs get redrawn on every click to the calendar. There should be a possibility of distinguishing between drag and drop motion and two clicks. If not, I would only keep the 2 click option.
> 
> Jun-10-2019 07-51-33
Opened task #459, this will include the "update data" button

**Comment by DzMakatun** [comment]
> ## 4 Make Records per run absolute values
> So if I have thousands of records I cannot see myself caring if I ingested 10K or 10K + 1 values. But I will care to see how the run does in comparison to others and the human eye cannot distinguish this if this is not put into absolute values.

Currently, the values are "number of records". Do you mean something different by "absolute values"? 
Or do you mean to make separate bar-charts for each status, instead of the stacked bars?  This could improve scaling for the smaller values, however,  the users will have to work with 6 charts where the most of the bars are empty. Also, the total number of records may vary between runs by a large factor. So the scaling problem will not disappear completely: 1 failed record will still not be visible comparing to 1000 failed records, while they both make 10% in different runs.

**Comment by DzMakatun** [comment]
> ## 6 Size of the column graph
> Column graph size greatly differs depending on the number of records in the selections. This could be unreadable with more records. I would set the record limit of ~20 (??) and in case there is more just show a warning, that there are more of them but showing the just the 20.
> 
> Or maybe let the user choose how many are displayed and when he displays 100 of them and it is hard to see, it is not our issue.

The number of runs displayed depends on the selected infodate interval. In task #459 I suggest to add "get data" button and add a warning before sending request if an interval larger than 1 month is selected.

**Comment by Zejnilovic** [comment]
> > ## 4 Make Records per run absolute values
> > So if I have thousands of records I cannot see myself caring if I ingested 10K or 10K + 1 values. But I will care to see how the run does in comparison to others and the human eye cannot distinguish this if this is not put into absolute values.
> 
> Currently, the values are "number of records". Do you mean something different by "absolute values"?
> Or do you mean to make separate bar-charts for each status, instead of the stacked bars? This could improve scaling for the smaller values, however, the users will have to work with 6 charts where the most of the bars are empty. Also, the total number of records may vary between runs by a large factor. So the scaling problem will not disappear completely: 1 failed record will still not be visible comparing to 1000 failed records, while they both make 10% in different runs.

Put a % on the X-axis is what I meant

File: menas/ui/components/external/it/designfuture/chartjs/library-preload.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (Zejnilovic) [reply menas/ui/components/external/it/designfuture/chartjs/library-preload.js]: But is this a way to use it? Can't we extract those string definitions of those modules?

**Comment by Zejnilovic** [comment]
> > ## 3 Summary date format
> > I would format the date in a more standard way of "%dd/%mm/%yyyy" for easier user experience.
> > image
> 
> The yyyy-MM-dd format matches the infodate (in the run collection, directory tree, job parameters, and chart labels). I would keep it this way, to be uniform across the monitoring data.

Uniformity is my goal as well. Info date comes from an _INFO file, which has a format of `dd-mm-yyyy` (my previous comment had `/` instead of `-`, that was my blunder. And for some reason `%`). 

All other places (`Run details` and `Records per run`) already do conform to this.

**Comment by Zejnilovic** [comment]
## 8 Formatting stage end time
Across our runs we always change the timestamp to conform to the time of the runner. Not showing the timezone. I think here it should be the same, so the user does not have to do the calculations.

image

image

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+      .aggregate(Seq(
+        // filter by dataset name
+        filter(equal("dataset", datasetName)),
+        //add casted dates
+        Document("""{$addFields: {
+                   |          startDateTimeCasted: {
+                   |              $dateFromString: {
+                   |                  dateString: "$startDateTime",
+                   |                  format: "%d-%m-%Y %H:%M:%S %z",
+                   |                  onError: {$dateFromString: {
```
[the remark this answers is not in the document]
Comment (DzMakatun) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala]: [062bda1](https://github.com/AbsaOSS/enceladus/pull/453/commits/062bda173c82c040e0faa3f46cda690d487ec31c) Switched to native mongodb scala driver methods where possible. It does not support needed Date and array expressions yet, so those parts had to remain. Revisited the tabs. The aggregation pipelines are used in this place only and will be a subject to change in the nearest future, so I would not split the query for now.

File: menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala (hunk 1 above)
Comment (DzMakatun) [reply menas/src/main/scala/za/co/absa/enceladus/menas/repositories/MonitoringMongoRepository.scala]: [062bda1](https://github.com/AbsaOSS/enceladus/pull/453/commits/062bda173c82c040e0faa3f46cda690d487ec31c) Switched to native mongodb scala driver methods where possible. It does not support needed Date and array expressions yet, so those parts had to remain.

File: menas/ui/service/MonitoringService.js (hunk 6 above)
Comment (DzMakatun) [reply menas/ui/service/MonitoringService.js]: Thnx for the heads up, this is a bigger one, created a new issue #473

**Comment by DzMakatun** [comment]
> ## 5 Button under headers
> The is a button under some headers and none under others. The button in general controls the view of the whole table, which makes it confusing when it is localized to one column, even more, when there are subcolumns named the same.
> 
> Jun-10-2019 10-56-28
> 
> Could this functionality be moved to some dropdown or something above the table?

Good point. I switch this off for now. Upon users's feedback the content of the table will be adjusted. I will come back to the table representation features afterwards.

**Comment by DzMakatun** [comment]
> > > ## 4 Make Records per run absolute values
> > > So if I have thousands of records I cannot see myself caring if I ingested 10K or 10K + 1 values. But I will care to see how the run does in comparison to others and the human eye cannot distinguish this if this is not put into absolute values.
> > 
> > 
> > Currently, the values are "number of records". Do you mean something different by "absolute values"?
> > Or do you mean to make separate bar-charts for each status, instead of the stacked bars? This could improve scaling for the smaller values, however, the users will have to work with 6 charts where the most of the bars are empty. Also, the total number of records may vary between runs by a large factor. So the scaling problem will not disappear completely: 1 failed record will still not be visible comparing to 1000 failed records, while they both make 10% in different runs.
> 
> Put a % on the X-axis is what I meant
ok, understood now, thnx. This and other ideas will be considered after the initial user feedback.

**Comment by DzMakatun** [comment]
> ## 8 Formatting stage end time
> Across our runs we always change the timestamp to conform to the time of the runner. Not showing the timezone. I think here it should be the same, so the user does not have to do the calculations.
> 
> image
> 
> image
> 
> Main point it to have it unified.

Added to issue #473, so that the Date and Timestamp issues will be addressed across  the monitoring all together.

File: menas/ui/components/external/it/designfuture/chartjs/library-preload.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (DzMakatun) [reply menas/ui/components/external/it/designfuture/chartjs/library-preload.js]: Not sure what you mean, let us discuss this later. This is the recommended ui5 way, as far as I understand...

**Comment by Zejnilovic** [comment]
BTW this branch currently has a partial merge and throws an error on login. But I believe this will be resolved after proper merge.

image

**Pull request merged**